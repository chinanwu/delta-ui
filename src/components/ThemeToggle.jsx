import PropTypes from 'prop-types';
import React, { useCallback } from 'react';
import { connect } from 'react-redux';

import { applyTheme } from '../thunk/ThemeThunk.jsx';

import './ThemeToggle.less';

export const ThemeToggle = ({ dark, onChangeTheme }) => {
	const handleChange = useCallback(() => {
		onChangeTheme(!dark);
	}, [dark, onChangeTheme]);

	return (
		<div className="ThemeToggle__container" aria-labelledby="themeToggleLabel">
			<label className="ThemeToggle">
				<input
					id="themeToggleInput"
					type="checkbox"
					checked={dark}
					aria-checked={dark}
					onChange={handleChange}
				/>
				<span className="ThemeToggle__slider" />
			</label>
			<div id="themeToggleLabel" className="ThemeToggle__label">
				Theme
			</div>
		</div>
	);
};

ThemeToggle.propTypes = {
	dark: PropTypes.bool,
	onChangeTheme: PropTypes.func,
};

export const mapStateToProps = ({ theme: { dark } }) => ({
	dark,
});

const mapDispatchToProps = {
	onChangeTheme: applyTheme,
};

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(ThemeToggle);
