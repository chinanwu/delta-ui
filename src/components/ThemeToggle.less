.ThemeToggle__container {
	position: absolute;
	top: 1rem;
	right: 1rem;
}

.ThemeToggle {
	position: relative;
	display: inline-block;
	width: 5rem;
	height: 2.6rem;
}

.ThemeToggle__label {
	position: relative;
	text-align: center;
}

.ThemeToggle input {
	opacity: 0;
	width: 0;
	height: 0;
}

.ThemeToggle__slider {
	position: absolute;
	cursor: pointer;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
	background-image: linear-gradient(
		to left top,
		#167aad,
		#1d8bbb,
		#299dc7,
		#38aed4,
		#48c0df
	);
	-webkit-transition: 0.4s;
	transition: 0.4s;
	border-radius: 3.4rem;
}

.ThemeToggle__slider:before {
	position: absolute;
	content: '';
	height: 1.8rem;
	width: 1.8rem;
	left: 0.4rem;
	bottom: 0.4rem;
	background-color: yellow;
	box-shadow: 0 0 12px 1px rgba(255, 247, 50, 0.53);
	-webkit-transition: 0.4s;
	transition: 0.4s;
	border-radius: 50%;
}

input:checked + .ThemeToggle__slider {
	background: rgb(64, 0, 87);
	background: linear-gradient(
		156deg,
		rgba(64, 0, 87, 1) 0%,
		rgba(25, 0, 92, 1) 35%,
		rgba(22, 26, 89, 1) 100%
	);

	border: 1px solid white;
}

input:checked + .ThemeToggle__slider:before {
	background-color: initial;
	box-shadow: none;
	background-image: url('../resources/moon.svg');
}

input:focus + .ThemeToggle__slider {
	box-shadow: 0 0 1px rgba(184, 175, 177, 0.25);
}

input:checked + .ThemeToggle__slider:before {
	-webkit-transform: translateX(2.4rem);
	-ms-transform: translateX(2.4rem);
	transform: translateX(2.4rem);
}
// From: https://www.w3schools.com/howto/howto_css_ThemeToggle.asp

@media only screen and (max-width: 375px) {
	.ThemeToggle__container {
		display: flex;
		flex-direction: column;
		align-items: end;
	}

	.ThemeToggle__label {
		padding-right: 0.4rem;
	}
}
