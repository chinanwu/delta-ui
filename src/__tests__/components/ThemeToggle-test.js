import React from 'react';
import { shallow } from 'enzyme';

import { ThemeToggle, mapStateToProps } from '../../components/ThemeToggle.jsx';

jest.unmock('../../components/ThemeToggle.jsx');

describe('ThemeToggle', () => {
	describe('rendering', () => {
		it('renders ThemeToggle', () => {
			const wrapper = shallow(<ThemeToggle />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders ThemeToggle dark mode', () => {
			const wrapper = shallow(
				<ThemeToggle dark={true} onChangeTheme={jest.fn()} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders ThemeToggle light', () => {
			const wrapper = shallow(
				<ThemeToggle dark={false} onChangeTheme={jest.fn()} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('changes the theme', () => {
			const onChangeTheme = jest.fn();
			const wrapper = shallow(
				<ThemeToggle dark={true} onChangeTheme={onChangeTheme} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
			wrapper.find('#themeToggleInput').simulate('change');
			expect(onChangeTheme.mock.calls).toEqual([[false]]);
		});
	});

	describe('mapStateToProps', () => {
		it('returns mapped properties', () => {
			const theme = { dark: true };
			expect(mapStateToProps({ theme })).toEqual({ dark: true });
		});
	});
});
