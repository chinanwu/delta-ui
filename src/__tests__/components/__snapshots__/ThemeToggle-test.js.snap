// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle behaviour changes the theme 1`] = `
<div
  className="ThemeToggle__container"
>
  <label
    className="ThemeToggle"
  >
    <input
      aria-checked={true}
      aria-labelledby="themeToggleLabel"
      checked={true}
      id="themeToggleInput"
      onChange={[Function]}
      type="checkbox"
    />
    <span />
  </label>
  <div
    className="ThemeToggle__label"
    id="themeToggleLabel"
  >
    Theme
  </div>
</div>
`;

exports[`ThemeToggle rendering renders ThemeToggle 1`] = `
<div
  className="ThemeToggle__container"
>
  <label
    className="ThemeToggle"
  >
    <input
      aria-labelledby="themeToggleLabel"
      id="themeToggleInput"
      onChange={[Function]}
      type="checkbox"
    />
    <span />
  </label>
  <div
    className="ThemeToggle__label"
    id="themeToggleLabel"
  >
    Theme
  </div>
</div>
`;

exports[`ThemeToggle rendering renders ThemeToggle dark mode 1`] = `
<div
  className="ThemeToggle__container"
>
  <label
    className="ThemeToggle"
  >
    <input
      aria-checked={true}
      aria-labelledby="themeToggleLabel"
      checked={true}
      id="themeToggleInput"
      onChange={[Function]}
      type="checkbox"
    />
    <span />
  </label>
  <div
    className="ThemeToggle__label"
    id="themeToggleLabel"
  >
    Theme
  </div>
</div>
`;

exports[`ThemeToggle rendering renders ThemeToggle light 1`] = `
<div
  className="ThemeToggle__container"
>
  <label
    className="ThemeToggle"
  >
    <input
      aria-checked={false}
      aria-labelledby="themeToggleLabel"
      checked={false}
      id="themeToggleInput"
      onChange={[Function]}
      type="checkbox"
    />
    <span />
  </label>
  <div
    className="ThemeToggle__label"
    id="themeToggleLabel"
  >
    Theme
  </div>
</div>
`;
